node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/Nodes.java

```diff
@@ -745,7 +745,7 @@ public List<Node> performOn(NodeList nodes, BiFunction<Node, Mutex, Node> action
 
     public List<Node> performOn(NodeList nodes, Predicate<Node> filter, BiFunction<Node, Mutex, Node> action) {
         List<Node> resultingNodes = new ArrayList<>();
-        nodes.stream().collect(groupingBy(Nodes::applicationIdForLock))
+        nodes.matching(filter).stream().collect(groupingBy(Nodes::applicationIdForLock))
```

Comment:
There are no guarantees as to when the filter is applied, so might as well apply it outside the lock. Whatever we read should be self-consisten.